### PR TITLE
Monomorphic flow nodes

### PR DESCRIPTION
**Pull request by ahejlsberg**
Experiment to observe the performance effects of monomorphic flow nodes.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test it

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this faster

**Comment by ahejlsberg** [comment]
@typescript-bot perf test this faster

**Comment by ahejlsberg** [comment]
@typescript-bot test it

**Comment by ahejlsberg** [comment]
@typescript-bot  perf test this faster

**Comment by jakebailey** [comment]
> Note that this is an API breaking change due to the more uniform layout of flow nodes.

To be honest, I'm not sure why any of `FlowNode` or `FlowFlags` are public; they are not referenced by anything but themselves. Makes me wonder if we should just be removing these so we can make more changes at will.

But, the API break will probably break the playground's flow node viewer and ts-ast-viewer.com's, but we can probably update both. Realistically the former should have been calling into `ts.Debug` anyway, and the latter pins specific TS versions, I think.

File: src/compiler/types.ts [hunk 1]
```diff
@@ -4101,19 +4102,28 @@ export type FlowNode =
 
 export interface FlowNodeBase {
     flags: FlowFlags;
-    id?: number; // Node id used by flow type cache in checker
+    id: number; // Node id used by flow type cache in checker
+    node: unknown; // Node or other data
+    antecedent: FlowNode | FlowNode[] | undefined;
```
Comment (jakebailey) [inline src/compiler/types.ts:4107]: I'm kind of surprised that this turned out okay; I would have thought these two types would have different shapes, but I guess they're both pointers...

Comment (ahejlsberg) [reply src/compiler/types.ts:4107]: They're both heap objects so should get the same hidden class.

**Comment by ahejlsberg** [comment]
@jakebailey Maybe we just fold the `@internal` change into this PR?

**Comment by jakebailey** [comment]
I figured it may be better not to, just to make it clear that we intentionally removed them, but I'm not against it if we're good with removing them.

If #58036 is merged first, then your PR doesn't have any breaks (no baseline changes at all).

**Comment by ahejlsberg** [comment]
> If https://github.com/microsoft/TypeScript/pull/58036 is merged first, then your PR doesn't have any breaks (no baseline changes at all).

Fine by me.

**Comment by jakebailey** [comment]
Done ☺️

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -496,9 +501,8 @@ export const enum ContainerFlags {
     IsObjectLiteralOrClassExpressionMethodOrAccessor = 1 << 7,
 }
 
-function initFlowNode<T extends FlowNode>(node: T) {
-    Debug.attachFlowNodeDebugInfo(node);
-    return node;
+export function createFlowNode(flags: FlowFlags, node: unknown, antecedent: FlowNode | FlowNode[] | undefined): FlowNode {
```
Comment (jakebailey) [inline src/compiler/binder.ts:504]:
```suggestion
/** @internal */
export function createFlowNode(flags: FlowFlags, node: unknown, antecedent: FlowNode | FlowNode[] | undefined): FlowNode {
```

Comment (ahejlsberg) [reply src/compiler/binder.ts:505]: Done!

**Review by jakebailey — APPROVED** [review]

**Pull request merged**